src/libcore/char.rs

```diff
@@ -676,3 +676,67 @@ impl Iterator for EncodeUtf16 {
         self.as_slice().iter().size_hint()
     }
 }
+
+
+/// An iterator over an iterator of bytes of the characters the bytes represent
+/// as UTF-8
+#[unstable(feature = "decode_utf8", issue = "33906")]
+#[derive(Clone, Debug)]
+pub struct DecodeUtf8<I: IntoIterator<Item = u8>>(DecodeUtf8Raw<I::IntoIter>);
+
+/// Decodes an `Iterator` of bytes as UTF-8.
+#[unstable(feature = "decode_utf8", issue = "33906")]
+#[inline]
+pub fn decode_utf8<I: IntoIterator<Item = u8>>(i: I) -> DecodeUtf8<I::IntoIter> {
+    DecodeUtf8(DecodeUtf8Raw(i.into_iter().peekable()))
+}
+
+/// `<DecodeUtf8 as Iterator>::next` returns this for an invalid input sequence.
+#[unstable(feature = "decode_utf8", issue = "33906")]
+#[derive(PartialEq, Debug)]
+pub enum InvalidSequence {
+    /// The byte sequence is not the shortest possible of the coded character.
+    TooLong(char, usize),
+    /// The code point is not a Unicode character.
+    BadCodePoint(u32),
+    /// The sequence is otherwise invalid.
+    Other,
+}
+
+#[unstable(feature = "decode_utf8", issue = "33906")]
+impl<I: Iterator<Item = u8>> Iterator for DecodeUtf8<I> {
+    type Item = Result<char, InvalidSequence>;
+    #[inline]
+    fn next(&mut self) -> Option<Result<char, InvalidSequence>> {
+        use self::InvalidSequence::*;
+        self.0.next().map(|r| r.map_err(|()| Other).and_then(|(x, l)| match from_u32(x) {
+            Some(x) => if l == x.len_utf8() { Ok(x) } else { Err(TooLong(x, l)) },
+            None => Err(BadCodePoint(x)),
+        }))
+    }
+}
+
+#[derive(Clone, Debug)]
+struct DecodeUtf8Raw<I: Iterator<Item = u8>>(::iter::Peekable<I>);
+
+impl<I: Iterator<Item = u8>> Iterator for DecodeUtf8Raw<I> {
+    type Item = Result<(u32, usize), ()>;
+    #[inline]
+    fn next(&mut self) -> Option<Result<(u32, usize), ()>> {
+        self.0.next().map(|b| if b & 0x80 == 0 { Ok((b as u32, 1)) } else {
```

Comment:
Stylistically we tend to have multi-line closures always have braces around them, putting this `if` on the next line.

You could make this similarly intended by doing:

```rust
foo.map(|b| {
    if condition {
        return Ok(...)
    }
    ...
})
```